### PR TITLE
Rewrite dump_tree method

### PR DESCRIPTION
**Pull request by eltimen**
Linked issues #1018 #359

* Increased performance (remove search elements in list before print it, reduces execution time from ~18 sec to ~2.2 for tree with len(dlg.descendants())=244)
* Added max children elements limit to print (`max_width` argument)

Also disabled output of best_match names for elements if max depth or width limit was reached, because determining best_match names for an element requires full tree traversal.

<details>
  <summary>Output example</summary>

  ```
  Warning: tree elements were printed up 2 depth level only. You can set a larger depth value or use depth=None to print a full tree (may freeze in case of very large number of elements)
Warning: determining best_match names for an element requires full tree traversal. Set depth and max_width parameters to None to print them.
Control Identifiers:

Dialog - 'Qt Main Window Example'    (L560, T280, R1360, B880)
child_window(name="Qt Main Window Example", control_type="Window")
   | 
   | TitleBar - ''    (L576, T258, R1360, B279)
   |    | 
   |    | Menu - 'Система'    (L560, T258, R581, B279)
   |    | child_window(name="Система", control_type="MenuBar")
   |    | 
   |    | Button - 'Свернуть'    (L1261, T260, R1292, B277)
   |    | child_window(name="Свернуть", control_type="Button")
   |    | 
   |    | Button - 'Развернуть'    (L1294, T260, R1325, B277)
   |    | child_window(name="Развернуть", control_type="Button")
   |    | 
   |    | **********
   |    | Max children output limit (3) has been reached. Set a larger max_width value or use max_width=None to see all children.
   |    | **********
   | 
   | Edit - ''    (L560, T343, R1360, B860)
   |    | 
   |    | GroupBox - ''    (L561, T344, R1359, B859)
   | 
   | Toolbar - 'Tool Bar 1'    (L560, T301, R818, B343)
   | child_window(name="Tool Bar 1", control_type="ToolBar")
   |    | 
   |    | Button - 'Two'    (L570, T304, R609, B342)
   |    | child_window(name="Two", control_type="Button")
   |    | 
   |    | Button - 'Four'    (L609, T304, R648, B342)
   |    | child_window(name="Four", control_type="Button")
   |    | 
   |    | Button - 'Three'    (L648, T304, R687, B342)
   |    | child_window(name="Three", control_type="Button")
   |    | 
   |    | **********
   |    | Max children output limit (3) has been reached. Set a larger max_width value or use max_width=None to see all children.
   |    | **********
   | 
   | **********
   | Max children output limit (3) has been reached. Set a larger max_width value or use max_width=None to see all children.
   | **********
  ```
</details>

### CODE REVIEW

File: pywinauto/base_application.py [hunk 1]
```diff
@@ -594,13 +595,19 @@ def _ctrl_identifiers(self):
...
 
-    def print_control_identifiers(self, depth=None, filename=None):
+    def print_control_identifiers(self, depth=30, max_width=5, filename=None):
         """
         Prints the 'identifiers'
 
         Prints identifiers for the control and for its descendants to
         a depth of **depth** (the whole subtree if **None**).
 
+        :param depth: Max depth level of an element tree to print (sNone - infinite).
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:605]: typo: `sNone` -> `None`

File: pywinauto/base_application.py [hunk 2]
```diff
@@ -594,13 +595,19 @@ def _ctrl_identifiers(self):
 
         return control_name_map
 
-    def print_control_identifiers(self, depth=None, filename=None):
+    def print_control_identifiers(self, depth=30, max_width=5, filename=None):
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:598]: Please set default `max_width=10` or maybe even 15-20. Also please check how much time it takes with `depth=30` for infinite tree case.

File: pywinauto/base_application.py [hunk 3]
```diff
@@ -594,13 +595,19 @@ def _ctrl_identifiers(self):
...
         """
         Prints the 'identifiers'
 
         Prints identifiers for the control and for its descendants to
         a depth of **depth** (the whole subtree if **None**).
 
+        :param depth: Max depth level of an element tree to print (sNone - infinite).
+
+        :param max_width: Max amount of children elements of each node to print (None - infinite)
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:607]: -> `Max number of children of each element to dump`

File: pywinauto/base_application.py [hunk 4]
```diff
@@ -594,13 +595,19 @@ def _ctrl_identifiers(self):
...
+
+        :param filename: Save tree to a specified file (None - print to console output)
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:609]: -> `(None: print to stdout)`

File: pywinauto/base_application.py (hunk 1 above)
Comment (eltimen) [reply pywinauto/base_application.py]: fixed

File: pywinauto/base_application.py (hunk 2 above)
Comment (eltimen) [reply pywinauto/base_application.py]: Default `max_width` and `depth` changed to 10.
Dolphin 5.0-9281 takes ~180s with `depth=30` and ~15-20s with `depth=10`

File: pywinauto/base_application.py (hunk 3 above)
Comment (eltimen) [reply pywinauto/base_application.py]: fixed

File: pywinauto/base_application.py (hunk 4 above)
Comment (eltimen) [reply pywinauto/base_application.py]: fixed

File: pywinauto/base_application.py [hunk 5]
```diff
@@ -609,49 +616,92 @@ def print_control_identifiers(self, depth=None, filename=None):
...
         # Wrap this control
         this_ctrl = self.__resolve_control(self.criteria)[-1]
 
-        # Create a list of this control and all its descendants
-        all_ctrls = [this_ctrl, ] + this_ctrl.descendants(depth=depth)
+        ElementTreeNode = collections.namedtuple('ElementTreeNode', ['elem', 'id', 'children'])
+
+        def create_element_tree(element_list):
+            """Build elements tree and create list with pre-order tree traversal"""
+            can_get_best_match_names = True
```
[the remark this answers is not in the document]
Comment (eltimen) [reply pywinauto/base_application.py]: fixed

File: pywinauto/base_application.py [hunk 6]
```diff
@@ -576,64 +577,117 @@ def _ctrl_identifiers(self):
...
             """Recursively print ids for ctrls and their descendants in a tree-like format"""
-            if len(ctrls) == 0 or current_depth > depth:
-                return
+            if current_depth == 0:
+                if depth_limit_reached:
+                    log_func('Warning: tree elements were printed up {} depth level only. '
+                             'You can set larger depth value or use depth=None to print full tree '
+                             '(may freeze in case of very large number of elements)'.format(depth))
+                if self.allow_magic_lookup and not show_best_match_names:
+                    log_func('Warning: determining best_match names for an element requires full tree traversal. '
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:670]: `determining` -> `calculating`

File: pywinauto/base_application.py [hunk 7]
```diff
@@ -576,64 +577,117 @@ def _ctrl_identifiers(self):
...
+                             'Set depth and max_width parameters to None to print them.')
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:671]: This line could be:
```
If the whole hierarchy fits into depth and max_width values, best_match names are dumped.
```

File: pywinauto/base_application.py [hunk 8]
```diff
@@ -576,64 +577,117 @@ def _ctrl_identifiers(self):
...
+            for name, index in name_ctrl_id_map.items():
+                ctrl_id_name_map.setdefault(index, []).append(name)
+
+        def print_identifiers(element_node, current_depth=0, log_func=print):
             """Recursively print ids for ctrls and their descendants in a tree-like format"""
-            if len(ctrls) == 0 or current_depth > depth:
-                return
+            if current_depth == 0:
+                if depth_limit_reached:
+                    log_func('Warning: tree elements were printed up {} depth level only. '
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:666]: `tree elements were printed up {} depth level only` -> `the whole hierarchy does not fit into depth={}. `

`You can set larger depth value or use depth=None to print full tree` -> `Increase depth parameter value or set to None (unlimited)`

File: pywinauto/base_application.py [hunk 9]
```diff
@@ -576,64 +577,117 @@ def _ctrl_identifiers(self):
...
+        Dump the 'identifiers' to console or a file
 
-        Prints identifiers for the control and for its descendants to
+        Dump identifiers for the control and for its descendants to
         a depth of **depth** (the whole subtree if **None**).
 
-        .. note:: The identifiers printed by this method have been made
+        :param depth: Max depth level of an element tree to dump (None: infinite).
+
+        :param max_width: Max number of children of each element to dump (None: infinite)
```
Comment (vasily-v-ryabov) [inline pywinauto/base_application.py:589]: `unlimited` sounds better than `infinite` in this context.

**Review by vasily-v-ryabov — APPROVED** [review]
Looks good. Just a little re-phrasing is needed.

File: pywinauto/base_application.py (hunk 7 above)
Comment (eltimen) [reply pywinauto/base_application.py]: changed

File: pywinauto/base_application.py (hunk 6 above)
Comment (eltimen) [reply pywinauto/base_application.py]: changed

File: pywinauto/base_application.py (hunk 8 above)
Comment (eltimen) [reply pywinauto/base_application.py]: changed

File: pywinauto/base_application.py (hunk 9 above)
Comment (eltimen) [reply pywinauto/base_application.py]: Yes, it looks better actually.

**Comment by vasily-v-ryabov** [comment]
@eltimen good job! Thanks!

**Pull request merged**